### ISSUE TITLE
amp-carousel: Best practice: Autoplaying carousels only stop/pause after interaction

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#29557]
## WCAG Level
Level A

## Priority
Medium

## Type
Interaction

## Pages/Components affected
* [amp-carousel](https://amp.dev/documentation/components/amp-carousel/)

## Description
Carousels are intended to present different blocks of content in a way that attracts attention and saves space, but if they are not implemented carefully they can have the opposite effect of pushing users away from the page entirely.

`<amp-carousel>` currently has the option to set the carousel to `autoplay`. By default, this cycles/advances the slides, with an author-defined `delay`. The only way to stop the carousel from auto-advancing is to interact with the previous/next buttons - once a user has explicitly activated one of these buttons, the autoplaying functionality is stopped. While this nominally passes WCAG 2.1 Success Criterion 2.2.2 Pause, Stop, Hide - as there is a way for users to stop the behaviour - this is not immediately obvious to users.

## User impact
Someone with a vestibular disorder may find that the constantly updating carousel makes them feel nauseous, whereas someone who finds reading difficult, perhaps because they have Dyslexia, will struggle to read the content of each block before it is replaced by the next. This issue can be even more problematic for assistive technology/screen reader users - if they are currently reading/interacting with content in the carousel, and the slide that they are currently on is advanced (hiding the slide and replacing it with new content), their focus may be lost/reset to the start of the page; even if their browser/assistive technology compensates for this by keeping their focus inside/near the carousel (rather than resetting it), the experience will be disorienting, as they may not realise the reason why the content they were currently on suddenly disappeared.

## Required solution
In general, authors should avoid carousels that update automatically, and instead rely solely on users explicitly changing slides with the previous/next controls.

If the carousel must update automatically, we recommend providing a "Play/Pause" control that stops the autoplaying behaviour. This can then be used explicitly by users, in addition to the current behaviour that pauses/stops the autoplaying when they use the previous/next buttons.

We further recommend pausing the carousel as soon as the user's mouse/pointer is over the carousel, and when any of the controls or content inside the carousel receive keyboard focus. This will make sure that content doesn't change "under the user's feet" as they navigate the page.

One additional suggestion would be to take advantage of the [`prefers-reduced-motion`](https://www.w3.org/TR/mediaqueries-5/#prefers-reduced-motion) media query. If the user has indicated a preference for reduced motion/animations, the autoplaying carousel could be paused by default (with the option for users to explicitly enable autoplay with the "Play/Pause" control if they wish).

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied:
1. Check that the carousel does not automatically update.
2. Check that there are controls to let users move backwards and forwards through the carousel.
3. If the carousel must update automatically, check there is a button to let users stop and restart it.
4. If the carousel must update automatically, check that it stops updating automatically once the user's mouse or keyboard focus are over/inside the carousel.

## Related standards
* [WCAG 2.1 Success Criterion 2.2.2 Pause, Stop, Hide (Level A)](https://www.w3.org/TR/WCAG21/#pause-stop-hide)

## More information
* [W3C - Understanding Success Criterion 2.2.2: Pause, Stop, Hide](https://www.w3.org/WAI/WCAG21/Understanding/pause-stop-hide.html)
* [W3C - Technique G186: Using a control in the Web page that stops moving, blinking, or auto-updating content](https://www.w3.org/WAI/WCAG21/Techniques/general/G186)
* [W3C - Technique G4: Allowing the content to be paused and restarted from where it was paused](https://www.w3.org/WAI/WCAG21/Techniques/general/G4)
* [W3C - Web Accessibility Tutorials: Carousel Concepts](https://www.w3.org/WAI/tutorials/carousels/)
* [Should I use a carousel?](http://shouldiuseacarousel.com/)
* [W3C Media Queries Level 5 / 12.2. Detecting the desire for less motion on the page: the `prefers-reduced-motion` feature](https://www.w3.org/TR/mediaqueries-5/#prefers-reduced-motion)
* [MDN - `prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
* [web.dev - prefers-reduced-motion: Sometimes less movement is more](https://web.dev/prefers-reduced-motion/)

## Test data
Test date: July 2020

**Comment by micajuine-ho** [comment]
@TetraLogicalHelpdesk Thank you for this very helpful analysis! I put this issue into the our Q4 a11y fix-it week sprint. 

> If the carousel must update automatically, we recommend providing a "Play/Pause" control that stops the autoplaying behaviour. This can then be used explicitly by users, in addition to the current behaviour that pauses/stops the autoplaying when they use the previous/next buttons.

This request seems to be a larger change and would require more planning, so I will not tackle this it in this sprint.

> We further recommend pausing the carousel as soon as the user's mouse/pointer is over the carousel, and when any of the controls or content inside the carousel receive keyboard focus. This will make sure that content doesn't change "under the user's feet" as they navigate the page.
>
> One additional suggestion would be to take advantage of the [`prefers-reduced-motion`](https://www.w3.org/TR/mediaqueries-5/#prefers-reduced-motion) media query. If the user has indicated a preference for reduced motion/animations, the autoplaying carousel could be paused by default (with the option for users to explicitly enable autoplay with the "Play/Pause" control if they wish).

I will address these two points in my upcoming PR.

**Comment by kristoferbaxter** [comment]
> One additional suggestion would be to take advantage of the prefers-reduced-motion media query. If the user has indicated a preference for reduced motion/animations, the autoplaying carousel could be paused by default (with the option for users to explicitly enable autoplay with the "Play/Pause" control if they wish).

Can we make this configurable for document developers as part of this change? This would be a breaking change in the eyes of many domains.

**Comment by micajuine-ho** [comment]
> Can we make this configurable for document developers as part of this change? This would be a breaking change in the eyes of many domains.

Happy to do this. However, I'm wondering, is it generally better to offload the accessibility responsibilities/configurations onto the publishers rather than being opinionated ourselves?

**Comment by kristoferbaxter** [comment]
> Happy to do this. However, I'm wondering, is it generally better to offload the accessibility responsibilities/configurations onto the publishers rather than being opinionated ourselves?

To make a breaking change to an existing component should be discussed at a Design Review, but in general avoided if possible.

The next major revision could introduce this by default though.

Also, I'd recommend bringing this topic to design review (in short format) so others can weigh in too!

**Comment by micajuine-ho** [comment]
> One additional suggestion would be to take advantage of the prefers-reduced-motion media query. If the user has indicated a preference for reduced motion/animations, the autoplaying carousel could be paused by default (with the option for users to explicitly enable autoplay with the "Play/Pause" control if they wish).

@kristoferbaxter  To clarify, are you suggesting that I should bring this whole issue to design review or just this feature?

**Comment by micajuine-ho** [comment]
> We further recommend pausing the carousel as soon as the user's mouse/pointer is over the carousel, and when any of the controls or content inside the carousel receive keyboard focus. This will make sure that content doesn't change "under the user's feet" as they navigate the page.

@TetraLogicalHelpdesk This change makes sense for pages not on mobile. Is there another recommendation for pages on mobile? 

My thought would be that if a user tapped/clicked on the image while on mobile, then we should also stop the autoplay.

**Comment by kristoferbaxter** [comment]
> @kristoferbaxter To clarify, are you suggesting that I should bring this whole issue to design review or just this feature?

Just the breaking change.

**Comment by TetraLogicalHelpdesk** [comment]
> > We further recommend pausing the carousel as soon as the user's mouse/pointer is over the carousel, and when any of the controls or content inside the carousel receive keyboard focus. This will make sure that content doesn't change "under the user's feet" as they navigate the page.
> 
> @TetraLogicalHelpdesk This change makes sense for pages not on mobile. Is there another recommendation for pages on mobile?

Conceptually, note that mouse/hover-capable inputs are not exclusive just to desktop/laptop - you can/will find users that have paired mice (or hovering stylus/pen inputs) even on mobile (see https://css-tricks.com/interaction-media-features-and-their-potential-for-incorrect-assumptions/#whats-the-primary-input) - so the hover-related approach should be taken for mobile as well. Ditto for keyboard use - users may have a bluetooth keyboard paired with their mobile/tablet and navigate using that. So the "if any control/element inside the carousel receives focus, stop the carousel" approach makes sense even in the "mobile" scenario.

> My thought would be that if a user tapped/clicked on the image while on mobile, then we should also stop the autoplay.

Agree, in the case of touch interactions, it makes sense to stop the autoplay as soon as the user has interacted for the first time with the carousel (either tapping it / one of its controls, or used a swipe gesture, or similar). You *may* decide that after a set period of inactivity, or when a tap/action is registered elsewhere outside of the carousel, autoplaying may be restarted (incidentally, this is what I implemented in Bootstrap a long time ago - see https://github.com/twbs/bootstrap/pull/22442). But that can still result in user annoyance, and be quite "quirky" in certain browsers, so if at all possible, the safest option is to stop autoplay and keep it stopped.
